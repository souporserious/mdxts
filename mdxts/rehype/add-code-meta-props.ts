import type { Element } from 'hast'
import type { VFile } from 'vfile'

/** Pass through meta string and code as props to `code` elements. */
export async function addCodeMetaProps(tree: Element, file: VFile) {
  const { visit } = await import('unist-util-visit')
  const { toString } = await import('hast-util-to-string')

  visit(tree, 'element', (element) => {
    if (element.tagName === 'pre') {
      const codeNode = element.children[0]

      if (
        codeNode &&
        codeNode.type === 'element' &&
        codeNode.tagName === 'code'
      ) {
        element.properties.code = toString(element)

        const meta = codeNode.data?.meta as string | undefined

        /* Map meta string to props. */
        meta?.split(' ').forEach((prop) => {
          const [key, value] = prop.split('=')
          element.properties[key] = value ?? true
        })
      }
    }
  })
}
