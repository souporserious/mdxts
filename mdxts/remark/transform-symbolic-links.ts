import type { Node, Parent, Text } from 'mdast'

/** Replace all symbolic links `[[link]]` with markdown links `[link](/link)`. */
export function transformSymbolicLinks() {
  return async (tree: Node) => {
    const { visitParents } = await import('unist-util-visit-parents')

    visitParents(tree, 'text', (node: Text, ancestors: Parent[]) => {
      const matches = node.value.match(/\[\[(.+?)\]\]/g)

      if (!matches) {
        return
      }

      const splitNodes: any[] = []
      let lastIndex = 0

      for (const match of matches) {
        const index = node.value.indexOf(match, lastIndex)
        const linkText = match.slice(2, -2)

        splitNodes.push({
          type: 'text',
          value: node.value.slice(lastIndex, index),
        })

        // Creating a markdown link instead of an HTML element
        splitNodes.push({
          type: 'link',
          url: `/${linkText}`,
          children: [{ type: 'text', value: linkText }],
        })

        lastIndex = index + match.length
      }

      splitNodes.push({
        type: 'text',
        value: node.value.slice(lastIndex),
      })

      const lastParent = ancestors[ancestors.length - 1]
      const startIndex = lastParent.children.indexOf(node)

      lastParent.children.splice(startIndex, 1, ...splitNodes)
    })
  }
}
