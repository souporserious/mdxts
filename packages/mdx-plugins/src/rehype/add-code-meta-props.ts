import type { Parent } from 'unist'
import type { Element, Properties } from 'hast'
import { toString } from 'hast-util-to-string'
import { visit, SKIP } from 'unist-util-visit'

interface CodeMetaElement extends Element {
  data?: {
    meta?: string
  }
  properties: Properties & {
    className?: string | string[]
    filename?: string
    language?: string
  }
}

/** Parses `CodeBlock` and `CodeInline` props and adds them to `pre` and `code` element properties respectively. */
export function addCodeMetaProps() {
  return (tree: Parent) => {
    visit(tree, 'element', (element: CodeMetaElement) => {
      if (element.tagName === 'pre') {
        const codeNode = element.children[0] as CodeMetaElement

        // Map meta string to props
        const meta = codeNode.data?.meta
        const props: Record<string, any> = {}

        meta?.split(' ').forEach((prop) => {
          const indexOfFirstEquals = prop.indexOf('=')
          if (indexOfFirstEquals === -1) {
            // Coerce boolean props to true if they don't have an explicit value
            props[prop] = true
          } else {
            const key = prop.substring(0, indexOfFirstEquals)
            const value = prop.substring(indexOfFirstEquals + 1)
            // Strip surrounding quotes if present
            props[key] = value.replace(/^["']|["']$/g, '')
          }
        })

        // Add props to code element
        Object.assign(element.properties, props)

        if (
          codeNode &&
          codeNode.type === 'element' &&
          codeNode.tagName === 'code'
        ) {
          const codeString = toString(codeNode)
          element.properties.value = codeString

          // get class name from code element
          const className = codeNode.properties.className as string
          const metadata = getClassNameMetadata(className || '')

          // Add filename and language as a props if they don't already exist
          if (metadata) {
            if (!element.properties.filename && metadata.filename) {
              element.properties.filename = metadata.filename
            }

            if (!element.properties.language) {
              element.properties.language = metadata.language
            }
          }
        }

        return SKIP
      } else if (element.tagName === 'code') {
        const codeString = toString(element)
        const firstSpaceIndex = codeString.indexOf(' ')

        if (firstSpaceIndex > -1) {
          const language = codeString.substring(0, firstSpaceIndex)

          // Add the language as a prop for syntax highlighting
          element.properties.language = language

          // Replace the element's content with just the code
          element.children = [
            {
              type: 'text',
              value: codeString.substring(firstSpaceIndex + 1),
            },
          ]
        }
      }
    })
  }
}

const languageKey = 'language-'
const languageLength = languageKey.length

/** Parses file metadata from a remark code block class name. */
function getClassNameMetadata(className: string | string[]) {
  const classNames = Array.isArray(className) ? className : className.split(' ')
  const filenameOrLanguage = classNames
    .find((name) => name.startsWith(languageKey))
    ?.slice(languageLength)

  if (!filenameOrLanguage) {
    return null
  }

  const extension = filenameOrLanguage.split('.').pop() ?? filenameOrLanguage

  return {
    filename: filenameOrLanguage?.includes('.') ? filenameOrLanguage : null,
    language: extension,
  } as {
    filename: string | null
    language: string
  }
}
