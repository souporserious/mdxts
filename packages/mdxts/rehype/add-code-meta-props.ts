import type { Element } from 'hast'
import type { VFile } from 'vfile'
import { transformCodeSync } from '../transform'
import { getLanguage } from './utils'

/** Pass through meta string and code as props to `code` elements. */
export async function addCodeMetaProps(tree: Element, file: VFile) {
  const { visit } = await import('unist-util-visit')
  const { toString } = await import('hast-util-to-string')

  visit(tree, 'element', (element) => {
    if (element.tagName === 'pre') {
      const codeNode = element.children[0]

      if (
        codeNode &&
        codeNode.type === 'element' &&
        codeNode.tagName === 'code'
      ) {
        const codeString = toString(element)
        const classNames = (codeNode.properties?.className || []) as string[]
        const language = getLanguage(classNames)

        element.properties.code = codeString

        /* Transform code block if marked as TypeScript or JavaScript. */
        if (/tsx?|jsx?/.test(language)) {
          try {
            element.properties.transformedCode = transformCodeSync(
              /* Wrap code in a function if it's inline JSX. */
              codeString.startsWith('<')
                ? `export default () => ${codeString}`
                : codeString
            )
          } catch (error) {
            console.error(
              `Error transforming MDX code block meta string for "${file.path}:${element.position?.start.line}"\n`,
              error
            )
          }
        }

        const meta = codeNode.data?.meta as string | undefined

        /* Map meta string to props. */
        meta?.split(' ').forEach((prop) => {
          const [key, value] = prop.split('=')
          element.properties[key] = value ?? true
        })
      }
    }
  })
}
