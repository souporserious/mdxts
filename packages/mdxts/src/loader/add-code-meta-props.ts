import type { JsxOpeningElement, JsxSelfClosingElement } from 'ts-morph'
import { Project, Node } from 'ts-morph'

import { getSourcePath } from '../utils/get-source-path'

/** Adds code meta props to MDXTS components. */
export function addCodeMetaProps(
  sourceCode: string,
  resourcePath: string,
  workingDirectory: string,
  gitSource?: string,
  gitBranch?: string,
  gitProvider?: string
): string {
  const project = new Project({ useInMemoryFileSystem: true })
  const sourceFile = project.createSourceFile('index.tsx', sourceCode)
  const jsxElements = sourceFile.getDescendants().filter((node) => {
    return Node.isJsxOpeningElement(node) || Node.isJsxSelfClosingElement(node)
  }) as (JsxOpeningElement | JsxSelfClosingElement)[]

  jsxElements.forEach((element) => {
    const tagName = element.getTagNameNode().getText()
    const attributesToAdd = []

    if (tagName === 'CodeBlock' || tagName === 'ExportedTypes') {
      const workingDirectoryAttribute = element.getAttribute('workingDirectory')
      if (!workingDirectoryAttribute) {
        attributesToAdd.push({
          name: 'workingDirectory',
          initializer: `"${workingDirectory}"`,
        })
      }
    }

    if (tagName === 'CodeBlock') {
      const sourcePathAttribute = element.getAttribute('sourcePath')
      if (!sourcePathAttribute) {
        const { line, column } = sourceFile.getLineAndColumnAtPos(
          element.getStart()
        )
        const sourcePath = getSourcePath(
          resourcePath,
          line,
          column,
          gitSource,
          gitBranch,
          gitProvider
        )
        attributesToAdd.push({
          name: 'sourcePath',
          initializer: `"${sourcePath}"`,
        })
      }
    }

    if (attributesToAdd.length > 0) {
      element.insertAttributes(0, attributesToAdd)
    }
  })

  return sourceFile.getFullText()
}
