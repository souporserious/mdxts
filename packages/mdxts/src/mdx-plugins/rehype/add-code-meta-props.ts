import type { Element, Root } from 'hast'
// import { BUNDLED_LANGUAGES } from 'shiki'

const languageMap: Record<string, any> = {
  mjs: 'javascript',
  js: 'javascript',
  jsx: 'javascript',
  ts: 'typescript',
  tsx: 'typescript',
}

const ADDITIONAL_LANGUAGES = Object.keys(languageMap)

/** Adds code meta props to the code element. */
export function addCodeMetaProps() {
  return async (tree: Root) => {
    const { visit } = await import('unist-util-visit')
    const { toString } = await import('hast-util-to-string')

    visit(tree, 'element', (element: Element) => {
      if (element.tagName === 'pre') {
        const codeNode = element.children[0]

        // Map meta string to props
        const meta = (codeNode.data as any)?.meta as string | undefined
        const props: Record<string, any> = {}
        meta?.split(' ').forEach((prop) => {
          const [key, value] = prop.split('=')
          props[key] =
            typeof value === 'undefined'
              ? true
              : value.replace(/^["']|["']$/g, '')
        })

        // Add props to code element
        Object.assign(element.properties, props)

        if (
          codeNode &&
          codeNode.type === 'element' &&
          codeNode.tagName === 'code'
        ) {
          const codeString = toString(codeNode)
          element.properties.code = codeString
        }
      } else if (element.tagName === 'code') {
        const codeString = toString(element)
        const firstSpaceIndex = codeString.indexOf(' ')

        if (firstSpaceIndex > -1) {
          const possibleLanguage = codeString.substring(0, firstSpaceIndex)
          const isValidLanguage = [{ id: '' }]
            .map((language) => language.id)
            .concat(ADDITIONAL_LANGUAGES)
            .includes(possibleLanguage)
          // const isValidLanguage = BUNDLED_LANGUAGES.map(
          //   (language) => language.id
          // )
          //   .concat(ADDITIONAL_LANGUAGES)
          //   .includes(possibleLanguage)

          if (isValidLanguage) {
            const language = languageMap[possibleLanguage] || possibleLanguage

            // Add the language as a prop for syntax highlighting
            element.properties.language = language

            // Replace the element's content with just the code
            element.children = [
              {
                type: 'text',
                value: codeString.substring(firstSpaceIndex + 1),
              },
            ]
          }
        }
      }
    })
  }
}
