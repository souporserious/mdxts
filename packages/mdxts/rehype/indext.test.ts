import { unified } from 'unified'
import rehypeParse from 'rehype-parse'
import rehypeStringify from 'rehype-stringify'
import { transformSymbolicLinks } from './transform-symbolic-links'
import { addCodeMetaProps } from './add-code-meta-props'

const symbolicLinksFixture = `Examples can be rendered in MDX using the helper [[Example]] and [[Preview]] components. These component take a \`source\` prop, which is the name of the example file or code block in any of the [locations examples can exist](./01.writing.mdx).`

test('transforms symbolic links correctly', async () => {
  const processor = unified()
    .use(rehypeParse)
    .use(() => transformSymbolicLinks)
    .use(rehypeStringify)
  const outputHtml = await processor.process(symbolicLinksFixture)

  expect(outputHtml).toMatchSnapshot()
})

const metaStringFixture = `
# Hello

\`\`\`tsx
const foo = 'bar'
\`\`\`
`

// TODO: Having ESM issues with this test.
test.skip('meta props passed through', async () => {
  const { compileSync } = await import('@mdx-js/mdx')
  const result = compileSync(metaStringFixture, {
    rehypePlugins: [() => addCodeMetaProps],
  })

  expect(result).toMatchSnapshot()
})
