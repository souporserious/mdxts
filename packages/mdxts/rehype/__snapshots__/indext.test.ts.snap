// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transforms symbolic links correctly 1`] = `
VFile {
  "cwd": "/Users/souporserious/Code/mdxts/mdxts",
  "data": {},
  "history": [],
  "messages": [],
  "value": "<html><head></head><body>Examples can be rendered in MDX using the helper <a href="/Example">Example</a> and <a href="/Preview">Preview</a> components. These component take a \`source\` prop, which is the name of the example file or code block in any of the [locations examples can exist](./01.writing.mdx).</body></html>",
}
`;
